lsp-completion.el

```diff
@@ -688,10 +688,15 @@ Others: CANDIDATES"
                                            'lsp-completion-item)
                              candidate
                            (cl-find candidate (funcall candidates) :test #'equal)))
+              (item (plist-get (text-properties-at 0 candidate) 'lsp-completion-item))
               (candidate
                ;; see #3498 typescript-language-server does not provide the
                ;; proper insertText without resolving.
-               (if (lsp-completion--find-workspace 'ts-ls)
+               (if (or (and (lsp-completion--find-workspace 'ts-ls)
+                            lsp-enable-snippet
+                            (eql (lsp:completion-item-insert-text-format? item)
+                                 lsp/insert-text-format-snippet))
+                       (eq (plist-get (lsp:completion-item-data? item) :hasAction) t))
```

Comment:
what's the `:hasAction` here? Don't use `plist-get` directly as it will not work for hashtable case. The `lsp:completion-item-data` if has a schema in `lsp-protocol` should expose this `hasAction` property as `:has-action`

Also, why increase the cases where we resolve the completion in blocking fashion? Before it only includes `ts-ls`, now it expanded to a new case as well.
The command and action if existed, should be resolved in non-blocking (async) way, see below where we resolve the completion item and apply command or additional snippets in non-blocking fashion. Since the resolved completion item is cached, the early return will be used.

Reply:
[CompletionItem.data](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem) in the protocol has an LSPAny type, it's arbitrary and has no schema. It's meant to be a catch all that every language server uses differently. The hasAction key is specific to ts-ls that I added in the PR I linked to.

> Don't use plist-get directly as it will not work for hashtable case

Ahhh, what else to use? Pretty much every function in this file pretends the hashtable case doesn't exist. I feel like `lsp-use-plists` should just be set to t at this point.

> Also, why increase the cases where we resolve the completion in blocking fashion? Before it only includes ts-ls, now it expanded to a new case as well.

brainfart, fixed.

Reply:
> Ahhh, what else to use?

There should be `lsp-get` or similar method in `lsp-protocol` that works for extracting property from both.

Reply:
Ohh I see what you mean. Fixed.

Reply:
> The command and action if existed, should be resolved in non-blocking (async) way, see below where we resolve the completion item and apply command or additional snippets in non-blocking fashion. Since the resolved completion item is cached, the early return will be used.

Ah thanks for pointing this out as well. I've deleted the hasAction condition